raster/src/main/scala/geotrellis/raster/io/geotiff/OverviewStrategy.scala

```diff
@@ -16,21 +16,78 @@
 
 package geotrellis.raster.io.geotiff
 
-sealed trait OverviewStrategy
+import geotrellis.raster.CellSize
+
+import scala.collection.Searching._
+
+/**
+  * Options ported from GDAL (https://gdal.org/programs/gdalwarp.html#cmdoption-gdalwarp-ovr)
+  *  for automatic selection from among available overviews
+  */
+sealed trait OverviewStrategy {
+  def selectOverview(overviewCS: List[CellSize], desiredCS: CellSize): CellSize
```

Comment:
Hm, we don't need this function in case of GDAL; do you think this is an appropriate place for selecting overview? 

This function is specific to the interpretation of the `OverviewStrategy` AST, and it is not a member of an AST that will work for all interpreters the same way. 

i.e. `GDALUtils.deriveOverviewStrategyString` lives in its own object, which is an interpreter for this AST.

What do you think about moving it in a separate object since it would be used only with GeoTrellis GeoTiffs?

Reply:
So far, the overview selection I've seen is pretty consistently be expressible in terms of the `selectOverview` method and this way of organizing things keeps the AST and the info necessary to make sense of it together + avoids duplication of some case-handling logic. I'm not opposed to our handling the overview selection case in a more general/appropriate fashion and therefore hiding this away for occasions where GDAL interop are required, but that seems like a different PR with a larger footprint.

Reply:
Hm, Is it correct to consider GDAL an "interop" thing? It could be an acceptable implementation if we were not using the same AST for GDAL. Moving all the specific to interpreter thing into interpreter looks pretty convenient btw (throwing here the RasterSources Query DSL). 

I'm happy to discuss it though.

Reply:
I think the point is somewhat academic and I'm not sure I'd have regrets in my life with either choice.

However, if I had to pick one I'd go with moving the `selectOverview` function to be on the companion object of`OverviewStrategy`. I have two soft reasons for feeling that way:

As @pomadchin mentions the purpose of this ADT is to select behavior. At the moment indication this selection to our GDAL bindings is "first class" intention of this ADT. Having a method that appears to do that on the instances of the ADT is slightly misleading because it could never be used by GDAL bindings.

This also reminds me of similar design choice made in [`ResampleTarget`](https://github.com/locationtech/geotrellis/blob/master/raster/src/main/scala/geotrellis/raster/ResampleTarget.scala) where the `apply` function does the job of creating the target `GridExtent`. I had to write some code that interpreted it in terms of old `Reproject.Options` and I found it hard to think through correctness there because trying to fit in the use of `apply` seemed it was a good thing. In the end I had to write interpretations of `ResampleTarget` over `Reproject.Options` that ignored the implementations of the `apply` method:

https://github.com/locationtech/geotrellis/blob/3e59f26807fc9d3a2dbf47fe358a1f1dbd2205ae/raster/src/main/scala/geotrellis/raster/ResampleTarget.scala#L87-L96

So in my mind that's the case where putting methods on ADT made it worse for interpreting the semantics between two similar domains for that ADT.

So perhaps the code reuse could be served by:

```scala
// before
val sourceCS = strategy.selectOverview(overviews.map(_.cellSize), cellSize)

// after
val sourceCS = OverviewStrategy.select(overviews.map(_.cellSize), cellSize, strategy)
```

Or even perhaps:
```scala
object OverviewStrategy {
   def select[A](overviews: A, desiredCS: CellSize, strategy: OverviewStrategy)(implicit ev: A => CellSize): A = {...}
}
```

Reply:
Reading between the lines I think this is going in the direction where we end up with two ways of selecting overviews: 

1. `GDALOverviewStrategy` which is basically what we have now
2. `OverviewStrategy` is that is JVM first in lines with #3194 (non-sealed with interface)

And there is a partial function `OverviewStrategy => GDALOverviewStrategy` that is used by `GDALRasterSource`.

Reply:
"Interop" might not be the right word choice In general, I like the interpreter pattern a lot but I'm not sure I see the appeal of separating out implementation from the 'language' (which in this case is just selection from among a few overview strategies) but here I don't fully see the appeal. I think that the AST's meaning is better documented to the extent that you can track down the code which will be called in case that option is supplied.

I think we might want more opinions to weigh in

Reply:
Having discussed this at length with @echeipesh and @pomadchin, I've removed `OverviewStrategy` methods in favor of a function on the companion object.